<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Hakon K. Gjessing and Julia Romanowska" />

<meta name="date" content="2019-04-17" />

<title>Reading and preparing the data</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Reading and preparing the data</h1>
<h4 class="author">Hakon K. Gjessing and Julia Romanowska</h4>
<h4 class="date">2019-04-17</h4>



<div id="reading-the-data-into-haplin" class="section level1">
<h1>Reading the data into Haplin</h1>
<p>Haplin reads data in two formats:</p>
<ol style="list-style-type: decimal">
<li><p>Haplin’s own text file format;</p></li>
<li><p>PED format.</p></li>
</ol>
<p>Both types of data are read in through the use of <code>genDataRead</code> function:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">dir.exmpl &lt;-<span class="st"> </span><span class="kw">system.file</span>( <span class="st">&quot;extdata&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;Haplin&quot;</span> )</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">exemplary.file1 &lt;-<span class="st"> </span><span class="kw">paste0</span>( dir.exmpl, <span class="st">&quot;/HAPLIN.trialdata.txt&quot;</span> )</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">my.gen.data.haplin &lt;-<span class="st"> </span><span class="kw">genDataRead</span>( <span class="dt">file.in =</span> exemplary.file1, <span class="dt">file.out =</span> <span class="st">&quot;trial_data1&quot;</span>,</a>
<a class="sourceLine" id="cb1-5" data-line-number="5">    <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">format =</span> <span class="st">&quot;haplin&quot;</span>, <span class="dt">n.vars =</span> <span class="dv">0</span> )</a></code></pre></div>
<pre><code>## Reading the data in chunks...
##  -- chunk 1 --
##  -- chunk 2 -- 
## ... done reading.
## Preparing data...
## ... done preparing
## Saving data...
## ... saved to files: ./trial_data1_gen.ffData, ./trial_data1_gen.RData</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">exemplary.file3 &lt;-<span class="st"> </span><span class="kw">paste0</span>( dir.exmpl, <span class="st">&quot;/exmpl_data.ped&quot;</span> )</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">my.gen.data &lt;-<span class="st"> </span><span class="kw">genDataRead</span>( exemplary.file3, <span class="dt">file.out =</span> <span class="st">&quot;ped_data&quot;</span>, <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span>,</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">    <span class="dt">format =</span> <span class="st">&quot;ped&quot;</span> )</a></code></pre></div>
<pre><code>## 'n.vars' was not given explicitly and will be set to 6 based on the format given.
## Reading the data in chunks...
##  -- chunk 1 --
##  -- chunk 2 -- 
## ... done reading.
## Preparing data...
## ... done preparing
## Saving data...
## ... saved to files: ./ped_data_gen.ffData, ./ped_data_gen.RData</code></pre>
<p>The function reads in all the data in the file, creates <code>ff</code> objects to store the genetic information and <code>data.frame</code> to store covariate data (if any). These objects are saved in <code>.RData</code> and <code>.ffData</code> files, which can be later on easily uploaded to R (with <code>genDataLoad</code>) and re-used.</p>
<p><em>CAUTION:</em> This can take a long time for large datasets (such as from GWAS analysis, e.g., reading in a 7 GB file will take ca.15 minutes), however, this needs to be run only once and then, the next time you need to use the data, use the <code>genDataLoad</code> function (see below). Be careful NOT TO DELETE the output files!</p>
<p>The <code>genDataRead</code> function returns a list object with two elements:</p>
<ul>
<li><em>cov.data</em> - a <code>data.frame</code> with covariate data (if available in the input file);</li>
<li><em>gen.data</em> - the genetic part of data;</li>
<li><em>aux</em> - additional, meta-information, used internal.</li>
</ul>
<p>The above function reads in also any additional covariate data the user has through the parameter <code>cov.file.in</code>:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">my.gen.data.haplin2 &lt;-<span class="st"> </span><span class="kw">genDataRead</span>( <span class="dt">file.in =</span> <span class="st">&quot;my_gen_data_hap.dat&quot;</span>,</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">  <span class="dt">file.out =</span> <span class="st">&quot;my_saved_gen_data_hap2&quot;</span>, <span class="dt">cov.file.in =</span> <span class="st">&quot;add_cov.dat&quot;</span>,</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">  <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">format =</span> <span class="st">&quot;haplin&quot;</span> )</a></code></pre></div>
<p>The file with the additional information should have a header with names of the data columns!</p>
<p>To see all the available arguments and usage examples, type:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">?genDataRead</a></code></pre></div>
<p>(this works also with any other function)</p>
<div id="reading-covariate-data" class="section level2">
<h2>Reading covariate data</h2>
<p>Together with the genotype data, the user can read covariates that are in a separate file. This is done at the same time as reading genotype data.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">add.cov.file &lt;-<span class="st"> </span><span class="kw">paste0</span>( dir.exmpl, <span class="st">&quot;/add_cov_data2.dat&quot;</span> )</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">my.gen.data.haplin3 &lt;-<span class="st"> </span><span class="kw">genDataRead</span>( <span class="dt">file.in =</span> exemplary.file1, <span class="dt">file.out =</span> <span class="st">&quot;trial_data3&quot;</span>,</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">    <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">format =</span> <span class="st">&quot;haplin&quot;</span>, <span class="dt">n.vars =</span> <span class="dv">0</span>, <span class="dt">cov.file.in =</span> add.cov.file )</a></code></pre></div>
<pre><code>## Reading the data in chunks...
##  -- chunk 1 --
##  -- chunk 2 -- 
## ... done reading.
## Preparing data...
## ... done preparing
## Reading covariate file... 
##     'cov.header' not given - assuming the first line is the header...
## ...done
## Saving data...
## ... saved to files: ./trial_data3_gen.ffData, ./trial_data3_gen.RData</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">my.gen.data.haplin3</a></code></pre></div>
<pre><code>## This is raw genetic data read in through genDataRead.
## 
## It contains the following parts:
##    cov.data, gen.data, aux 
## 
## with following dimensions:
##   - covariate variables =  smoke, vitamin, pheno1, pheno2 
##       (total  4  covariate variables),
##   - number of markers =  2 ,  - number of data lines =  249</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">add.cov.file2 &lt;-<span class="st"> </span><span class="kw">paste0</span>( dir.exmpl, <span class="st">&quot;/add_cov_data.dat&quot;</span> )</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">my.gen.data2 &lt;-<span class="st"> </span><span class="kw">genDataRead</span>( exemplary.file3, <span class="dt">file.out =</span> <span class="st">&quot;ped_data2&quot;</span>, <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span>,</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">    <span class="dt">format =</span> <span class="st">&quot;ped&quot;</span>, <span class="dt">cov.file.in =</span> add.cov.file2 )</a></code></pre></div>
<pre><code>## 'n.vars' was not given explicitly and will be set to 6 based on the format given.
## Reading the data in chunks...
##  -- chunk 1 --
##  -- chunk 2 -- 
## ... done reading.
## Preparing data...
## ... done preparing
## Reading covariate file... 
##     'cov.header' not given - assuming the first line is the header...
## ...done
## Saving data...
## ... saved to files: ./ped_data2_gen.ffData, ./ped_data2_gen.RData</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">my.gen.data2</a></code></pre></div>
<pre><code>## This is raw genetic data read in through genDataRead.
## 
## It contains the following parts:
##    cov.data, gen.data, aux 
## 
## with following dimensions:
##   - covariate variables =  id.fam, id.c, id.f, id.m, sex, cc, smoke, vitamin, pheno1, pheno2 
##       (total  10  covariate variables),
##   - number of markers =  429 ,  - number of data lines =  1659</code></pre>
</div>
</div>
<div id="accessing-the-information-in-the-loaded-data" class="section level1">
<h1>Accessing the information in the loaded data</h1>
<p>The object created by <code>genDataRead</code> includes a lot of information. We have created functions that will help the user to navigate it.</p>
<div id="displaying-and-extracting-phenotype-information" class="section level2">
<h2>Displaying and extracting phenotype information</h2>
<p>First of all, when you type in the name of the object, a short summary will be displayed:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">my.gen.data</a></code></pre></div>
<pre><code>## This is raw genetic data read in through genDataRead.
## 
## It contains the following parts:
##    cov.data, gen.data, aux 
## 
## with following dimensions:
##   - covariate variables =  id.fam, id.c, id.f, id.m, sex, cc 
##       (total  6  covariate variables),
##   - number of markers =  429 ,  - number of data lines =  1659</code></pre>
<p>If you want to show and/or extract part of phenotype information, you can use the <code>showPheno</code> function:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw">showPheno</span>( my.gen.data ) <span class="co"># by default - showing first 5 entries</span></a></code></pre></div>
<pre><code>##      id.fam id.c id.f id.m sex cc 
## [1,] &quot;1&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [2,] &quot;1&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [3,] &quot;1&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;
## [4,] &quot;2&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;0&quot;
## [5,] &quot;2&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw">head</span>( <span class="kw">showPheno</span>( my.gen.data, <span class="dt">n =</span> <span class="st">&quot;all&quot;</span> ), <span class="dt">n =</span> <span class="dv">20</span> )</a></code></pre></div>
<pre><code>##       id.fam id.c id.f id.m sex cc 
##  [1,] &quot;1&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
##  [2,] &quot;1&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [3,] &quot;1&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;
##  [4,] &quot;2&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;0&quot;
##  [5,] &quot;2&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [6,] &quot;2&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;
##  [7,] &quot;3&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;0&quot;
##  [8,] &quot;3&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
##  [9,] &quot;3&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;0&quot;
## [10,] &quot;4&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;1&quot;
## [11,] &quot;4&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [12,] &quot;4&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;0&quot;
## [13,] &quot;5&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [14,] &quot;5&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [15,] &quot;5&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;
## [16,] &quot;6&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;1&quot;
## [17,] &quot;6&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [18,] &quot;6&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;0&quot;
## [19,] &quot;7&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [20,] &quot;7&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="kw">showPheno</span>( my.gen.data, <span class="dt">from =</span> <span class="dv">4</span>, <span class="dt">to =</span> <span class="dv">15</span> )</a></code></pre></div>
<pre><code>##       id.fam id.c id.f id.m sex cc 
##  [1,] &quot;2&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;0&quot;
##  [2,] &quot;2&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [3,] &quot;2&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;
##  [4,] &quot;3&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;0&quot;
##  [5,] &quot;3&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
##  [6,] &quot;3&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;0&quot;
##  [7,] &quot;4&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;1&quot;
##  [8,] &quot;4&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [9,] &quot;4&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;0&quot;
## [10,] &quot;5&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [11,] &quot;5&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [12,] &quot;5&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw">head</span>( <span class="kw">showPheno</span>( my.gen.data, <span class="dt">sex =</span> <span class="dv">2</span> ), <span class="dt">n =</span> <span class="dv">20</span> ) <span class="co"># show information about females only</span></a></code></pre></div>
<pre><code>##       id.fam id.c id.f id.m sex cc 
##  [1,] &quot;1&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
##  [2,] &quot;1&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [3,] &quot;2&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [4,] &quot;3&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
##  [5,] &quot;4&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [6,] &quot;5&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
##  [7,] &quot;5&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [8,] &quot;6&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
##  [9,] &quot;7&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [10,] &quot;7&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [11,] &quot;8&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;0&quot;
## [12,] &quot;8&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [13,] &quot;9&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [14,] &quot;9&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [15,] &quot;10&quot;   &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [16,] &quot;11&quot;   &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [17,] &quot;11&quot;   &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [18,] &quot;12&quot;   &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [19,] &quot;13&quot;   &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [20,] &quot;14&quot;   &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;</code></pre>
<p>The output can be saved to an object:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1">females.pheno &lt;-<span class="st"> </span><span class="kw">showPheno</span>( my.gen.data, <span class="dt">sex =</span> <span class="dv">2</span> )</a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="kw">head</span>( females.pheno )</a></code></pre></div>
<pre><code>##      id.fam id.c id.f id.m sex cc 
## [1,] &quot;1&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [2,] &quot;1&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [3,] &quot;2&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [4,] &quot;3&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [5,] &quot;4&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [6,] &quot;5&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;</code></pre>
<p>With the functions <code>nindiv</code> and <code>nfam</code>, you can get the number of individuals or number of families in your data:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw">nindiv</span>( my.gen.data )</a></code></pre></div>
<pre><code>## [1] 1659</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="kw">nfam</span>( my.gen.data )</a></code></pre></div>
<pre><code>## [1] 550</code></pre>
<p>Note that the <code>nfam</code> function assumes that your dataset is from a triad study, i.e., includes at least child and one parent information.</p>
</div>
<div id="displaying-and-extracting-genotype-information" class="section level2">
<h2>Displaying and extracting genotype information</h2>
<p>The function <code>nsnps</code> will tell us how many markers/SNPs there is in the data. Be careful, as a default it assumes that the data is triad data (i.e., mother, father and child were genotyped), so if your data is from a case-control study, be sure to specify that with argument <code>design = &quot;cc&quot;</code>.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1"><span class="kw">nsnps</span>( my.gen.data )</a></code></pre></div>
<pre><code>## [1] 429</code></pre>
<p>To extract and/or show genotypes for specific individuals or markers, use the <code>showGen</code> function:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw">showGen</span>( my.gen.data, <span class="dt">markers =</span> <span class="kw">c</span>( <span class="dv">10</span>,<span class="dv">15</span>,<span class="dv">121</span> ) ) <span class="co"># by default - showing first 5 entries</span></a></code></pre></div>
<pre><code>## opening ff C:/Users/hkgjess/AppData/Local/Temp/RtmpecqmZK/ff3c905b8c7c23.ff</code></pre>
<pre><code>## ff (open) byte length=30 (30) dim=c(5,6) dimorder=c(1,2) levels: G &lt;NA&gt; C T A
##      [,1] [,2] [,3] [,4] [,5] [,6]
## [1,]    T    T    A    T    T    T
## [2,]    T    T    T    T    T    T
## [3,]    T    T    A    T    T    T
## [4,]    A    T    A    T    T    T
## [5,]    A    T    T    T    T    T</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="kw">showGen</span>( my.gen.data, <span class="dt">from =</span> <span class="dv">31</span>, <span class="dt">to =</span> <span class="dv">231</span> )</a></code></pre></div>
<pre><code>## ff (open) byte length=2010 (2010) dim=c(201,10) dimorder=c(1,2) levels: G &lt;NA&gt; C T A
##        [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]      G    G    T    A    T    T    G    G    G     G
## [2,]      G    G    A    A    T    T    G    G    G     G
## [3,]      G    G    T    A    T    T    G    G    G     G
## [4,]      G    G    A    A    A    T    G    G    C     G
## [5,]      G    G    T    A    A    T    G    G    C     G
## [6,]      G    G    A    A    A    T    G    G    C     G
## [7,]      G    G    A    A    A    T    G    G    G     G
## [8,]      G    G    A    A    A    T    G    G    G     G
## :         :    :    :    :    :    :    :    :    :     :
## [194,]    G    G    A    A    A    A    G    G    C     G
## [195,]    G    G    A    A    A    A    G    G    G     G
## [196,]    G    G    T    A    A    T    G    G    C     C
## [197,]    G    G    T    A    A    T    G    G    G     G
## [198,]    G    G    T    A    A    T    G    G    G     G
## [199,]    G    G    A    A    T    T    G    G    G     G
## [200,]    C    G    A    A    T    T    G    G    G     G
## [201,]    C    G    A    A    A    T    G    G    C     G</code></pre>
<p>As above, this output can be saved to an object:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1">subset.genes &lt;-<span class="st"> </span><span class="kw">showGen</span>( my.gen.data, <span class="dt">from =</span> <span class="dv">31</span>, <span class="dt">to =</span> <span class="dv">231</span>, <span class="dt">markers =</span> <span class="kw">c</span>( <span class="dv">10</span>,<span class="dv">15</span>,<span class="dv">121</span> ) )</a>
<a class="sourceLine" id="cb38-2" data-line-number="2">subset.genes</a></code></pre></div>
<pre><code>## ff (open) byte length=1206 (1206) dim=c(201,6) dimorder=c(1,2) levels: G &lt;NA&gt; C T A
##        [,1] [,2] [,3] [,4] [,5] [,6]
## [1,]     A    T    T    T    A    T 
## [2,]     A    T    T    T    T    T 
## [3,]     A    A    T    T    A    T 
## [4,]     T    T    T    T    A    T 
## [5,]     A    T    T    T    A    T 
## [6,]     T    T    T    T    A    T 
## [7,]     A    T    T    T    T    T 
## [8,]     A    T    T    T    T    T 
## :         :    :    :    :    :    :
## [194,]   A    T    T    T    T    T 
## [195,]   A    A    T    T    T    T 
## [196,]   T    T    T    T    T    T 
## [197,]   A    T    T    T    NA   NA
## [198,]   A    T    T    T    T    T 
## [199,]   A    A    T    T    A    T 
## [200,]   A    A    T    T    T    T 
## [201,]   A    A    T    T    T    T</code></pre>
<p>Note that these functions work only with objects resulting from <code>genDataRead</code>, and not <code>genDataPreprocess</code>, as the preprocessing disturbs the coding of the data, thus making the output not easy to understand.</p>
</div>
</div>
<div id="preparing-your-data" class="section level1">
<h1>Preparing your data</h1>
<p>After loading the data, it is necessary to pre-process it to the internal format used by haplin. This is done by evoking the command:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1">my.prepared.gen.data &lt;-<span class="st"> </span><span class="kw">genDataPreprocess</span>( <span class="dt">data.in =</span> my.gen.data, <span class="dt">map.file =</span></a>
<a class="sourceLine" id="cb40-2" data-line-number="2">  <span class="st">&quot;my_gen_data.map&quot;</span>, <span class="dt">design =</span> <span class="st">&quot;triad&quot;</span>, <span class="dt">file.out =</span> <span class="st">&quot;my_prepared_gen_data&quot;</span>,</a>
<a class="sourceLine" id="cb40-3" data-line-number="3">  <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span> )</a></code></pre></div>
<p><em>CAUTION:</em> This action can be very time-consuming for large datasets (e.g., estimated time for ca.45,000 SNPs and 1,600 individuals, a PED file of ca.700MB, is ca.6 minutes on a 7-core CPU). However, this needs to be done only once and the output, stored in small files, can be used for the subsequent analysis repeatedly. (See also section <a href="#subset">Choosing a subset of data</a></p>
<p>This will also create <code>.RData</code> and <code>.ffData</code> files, which take much less space than the input PED files. Be careful not to delete these files, as they can be re-used by simply loading into R (the `genDataLoad function) right before Haplin analysis.</p>
<p><em>NOTE:</em> The information on the <code>my.prepared.gen.data</code> object can be displayed by simply writing the name of the object.</p>
</div>
<div id="subset" class="section level1">
<h1>Choosing a subset of data</h1>
<p>If you know that you want to focus your analysis on a certain region of the entire SNP set, or perhaps you’re impatient and want to check out Haplin without waiting a long time for the preprocessing to finish, you can easily choose a subset of the data to pre-process and analyze. This can be done with the command:</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1">gen.data.subset &lt;-<span class="st"> </span><span class="kw">genDataGetPart</span>( <span class="dt">data.in =</span> my.gen.data, <span class="dt">markers =</span> <span class="kw">c</span>( <span class="dv">3</span><span class="op">:</span><span class="dv">15</span>,<span class="dv">22</span> ),</a>
<a class="sourceLine" id="cb41-2" data-line-number="2">  <span class="dt">design =</span> <span class="st">&quot;triad&quot;</span>, <span class="dt">file.out =</span> <span class="st">&quot;my_gen_data_subset&quot;</span>, <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span> )</a></code></pre></div>
<p>This function allows you to specify the subset in various ways:</p>
<ul>
<li><strong>markers</strong> - numeric vector with numbers indicating which markers to choose;</li>
<li><strong>indiv.ids</strong> - numeric vector giving IDs of individuals. <em>CAUTION:</em> in a standard PED file, individual IDs are not unique, so this will select all individuals with given numbers;</li>
<li><strong>rows</strong> - numeric vector giving the positions - this will select only these rows;</li>
<li><strong>cc</strong> - one or more values to choose based on the case-control status (‘cc’ column);</li>
<li><strong>sex</strong> - one or more values to choose based on the ‘sex’ column;</li>
<li><strong>…</strong> - if any additional covariate data are available in `data.in, the user can choose based on values of these.</li>
</ul>
<p>If you give a combination of these parameters, the result will be the intersection of the subsets defined by each of the parameters alone. The subset is then available in the <code>gen.data.subset</code> object and written to <code>file.out</code> file(s). These can be loaded and re-used multiple times.</p>
</div>
<div id="re-using-the-data" class="section level1">
<h1>Re-using the data</h1>
<p><em>IMPORTANT:</em> Remember that each time you start a new R session, you need to load the data into the memory with the command:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1">my.prepared.gen.data &lt;-<span class="st"> </span><span class="kw">genDataLoad</span>( <span class="dt">filename =</span> <span class="st">&quot;my_prepared_gen_data&quot;</span>,</a>
<a class="sourceLine" id="cb42-2" data-line-number="2">  <span class="dt">dir.in =</span> <span class="st">&quot;.&quot;</span> )</a></code></pre></div>
<p>This takes much less time than re-reading and running the data preparations! The output of the <code>genDataPreprocess</code> function (or <code>genDataLoad</code>) can then be used to run the analysis.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
