<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Hakon K. Gjessing and Julia Romanowska" />

<meta name="date" content="2020-09-05" />

<title>Reading and preparing the data</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Reading and preparing the data</h1>
<h4 class="author">Hakon K. Gjessing and Julia Romanowska</h4>
<h4 class="date">2020-09-05</h4>



<div id="reading-the-data-into-haplin" class="section level1">
<h1>Reading the data into Haplin</h1>
<p>Haplin reads data in two formats:</p>
<ol style="list-style-type: decimal">
<li><p>Haplinâ€™s own text file format;</p></li>
<li><p>PED format (as generated by <code>plink 1.9</code>).</p></li>
</ol>
<p>Both types of data are read in through the use of <code>genDataRead</code> function:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>dir.exmpl &lt;-<span class="st"> </span><span class="kw">system.file</span>( <span class="st">&quot;extdata&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;Haplin&quot;</span> )</span>
<span id="cb1-2"><a href="#cb1-2"></a>exemplary.file1 &lt;-<span class="st"> </span><span class="kw">paste0</span>( dir.exmpl, <span class="st">&quot;/HAPLIN.trialdata.txt&quot;</span> )</span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a>my.gen.data.haplin &lt;-<span class="st"> </span><span class="kw">genDataRead</span>( <span class="dt">file.in =</span> exemplary.file1, <span class="dt">file.out =</span> <span class="st">&quot;trial_data1&quot;</span>,</span>
<span id="cb1-5"><a href="#cb1-5"></a>    <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">format =</span> <span class="st">&quot;haplin&quot;</span>, <span class="dt">n.vars =</span> <span class="dv">0</span> )</span></code></pre></div>
<pre><code>## Reading the data in chunks...
##  -- chunk 1 --
##  -- chunk 2 -- 
## ... done reading.
## Reading the marker names...</code></pre>
<pre><code>## Warning: No map file given, map file empty or the number of map file rows not
## equal to the number of markers in data; will generate dummy marker names.</code></pre>
<pre><code>## ...done
## Preparing data...
## ... done preparing
## Saving data...
## ... saved to files: ./trial_data1_gen.ffData, ./trial_data1_gen.RData</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>exemplary.file3 &lt;-<span class="st"> </span><span class="kw">paste0</span>( dir.exmpl, <span class="st">&quot;/exmpl_data.ped&quot;</span> )</span>
<span id="cb5-2"><a href="#cb5-2"></a>my.gen.data &lt;-<span class="st"> </span><span class="kw">genDataRead</span>( exemplary.file3, <span class="dt">file.out =</span> <span class="st">&quot;ped_data&quot;</span>, <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span>,</span>
<span id="cb5-3"><a href="#cb5-3"></a>    <span class="dt">format =</span> <span class="st">&quot;ped&quot;</span> )</span></code></pre></div>
<pre><code>## &#39;n.vars&#39; was not given explicitly and will be set to 6 based on the format given.
## Reading the data in chunks...
##  -- chunk 1 --
##  -- chunk 2 -- 
## ... done reading.
## Reading the marker names...</code></pre>
<pre><code>## Warning: No map file given, map file empty or the number of map file rows not
## equal to the number of markers in data; will generate dummy marker names.</code></pre>
<pre><code>## ...done
## Preparing data...
## ... done preparing
## Saving data...
## ... saved to files: ./ped_data_gen.ffData, ./ped_data_gen.RData</code></pre>
<p>The function reads in all the data in the file, creates <code>ff</code> objects to store the genetic information and <code>data.frame</code> to store covariate data (if any). These objects are saved in <code>.RData</code> and <code>.ffData</code> files, which can be later on easily uploaded to R (with <code>genDataLoad</code>) and re-used.</p>
<p><em>CAUTION:</em> This can take a long time for large datasets (such as from GWAS analysis, e.g., reading in a 7 GB file will take ca.15 minutes), however, this needs to be run only once and then, the next time you need to use the data, use the <code>genDataLoad</code> function (see section <a href="#reuse">Re-using the data</a>, below). Be careful NOT TO DELETE the output files .ffData and .RData!</p>
<p>The <code>genDataRead</code> function returns a list object with three elements:</p>
<ul>
<li><em>cov.data</em> - a <code>data.frame</code> with covariate data (if available in the input file);</li>
<li><em>gen.data</em> - the genetic part of data;</li>
<li><em>aux</em> - additional meta-information, used internal.</li>
</ul>
<p>To see all the available arguments and usage examples, type:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>?genDataRead</span>
<span id="cb9-2"><a href="#cb9-2"></a></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="kw">example</span>( genDataRead )</span></code></pre></div>
<p>(this works also with any other function)</p>
<div id="reading-covariate-data" class="section level2">
<h2>Reading covariate data</h2>
<p>The PED-formatted data has by default 6 first columns reserved for family information and covariates. However, if the user has covariate data that are in a separate file, this can be read together with the genotype data, using the same function <code>genDataRead</code>. The main file can be either in haplin or PED format.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>add.cov.file &lt;-<span class="st"> </span><span class="kw">paste0</span>( dir.exmpl, <span class="st">&quot;/add_cov_data2.dat&quot;</span> )</span>
<span id="cb10-2"><a href="#cb10-2"></a>my.gen.data.haplin3 &lt;-<span class="st"> </span><span class="kw">genDataRead</span>( <span class="dt">file.in =</span> exemplary.file1, <span class="dt">file.out =</span> <span class="st">&quot;trial_data3&quot;</span>,</span>
<span id="cb10-3"><a href="#cb10-3"></a>    <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">format =</span> <span class="st">&quot;haplin&quot;</span>, <span class="dt">n.vars =</span> <span class="dv">0</span>, <span class="dt">cov.file.in =</span> add.cov.file )</span></code></pre></div>
<pre><code>## Reading the data in chunks...
##  -- chunk 1 --
##  -- chunk 2 -- 
## ... done reading.
## Reading the marker names...</code></pre>
<pre><code>## Warning: No map file given, map file empty or the number of map file rows not
## equal to the number of markers in data; will generate dummy marker names.</code></pre>
<pre><code>## ...done
## Preparing data...
## ... done preparing
## Reading covariate file... 
##     &#39;cov.header&#39; not given - assuming the first line is the header...
## ...done
## Saving data...
## ... saved to files: ./trial_data3_gen.ffData, ./trial_data3_gen.RData</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>my.gen.data.haplin3</span></code></pre></div>
<pre><code>## This is raw genetic data read in through genDataRead.
## 
## It contains the following parts:
##    cov.data, gen.data, aux 
## 
## with following dimensions:
##   - covariate variables =  smoke, vitamin, pheno1, pheno2 
##       (total  4  covariate variables),
##   - number of markers =  2 ,
##   - number of data lines =  249</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>add.cov.file2 &lt;-<span class="st"> </span><span class="kw">paste0</span>( dir.exmpl, <span class="st">&quot;/add_cov_data.dat&quot;</span> )</span>
<span id="cb16-2"><a href="#cb16-2"></a>my.gen.data2 &lt;-<span class="st"> </span><span class="kw">genDataRead</span>( exemplary.file3, <span class="dt">file.out =</span> <span class="st">&quot;ped_data2&quot;</span>, <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span>,</span>
<span id="cb16-3"><a href="#cb16-3"></a>    <span class="dt">format =</span> <span class="st">&quot;ped&quot;</span>, <span class="dt">cov.file.in =</span> add.cov.file2 )</span></code></pre></div>
<pre><code>## &#39;n.vars&#39; was not given explicitly and will be set to 6 based on the format given.
## Reading the data in chunks...
##  -- chunk 1 --
##  -- chunk 2 -- 
## ... done reading.
## Reading the marker names...</code></pre>
<pre><code>## Warning: No map file given, map file empty or the number of map file rows not
## equal to the number of markers in data; will generate dummy marker names.</code></pre>
<pre><code>## ...done
## Preparing data...
## ... done preparing
## Reading covariate file... 
##     &#39;cov.header&#39; not given - assuming the first line is the header...
## ...done
## Saving data...
## ... saved to files: ./ped_data2_gen.ffData, ./ped_data2_gen.RData</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>my.gen.data2</span></code></pre></div>
<pre><code>## This is raw genetic data read in through genDataRead.
## 
## It contains the following parts:
##    cov.data, gen.data, aux 
## 
## with following dimensions:
##   - covariate variables =  id.fam, id.c, id.f, id.m, sex, cc, smoke, vitamin, pheno1, pheno2 
##       (total  10  covariate variables),
##   - number of markers =  429 ,
##   - number of data lines =  1659</code></pre>
<p><em>NOTE:</em> The file with the additional information should have a header with names of the data columns!</p>
</div>
</div>
<div id="accessing-the-information-in-the-loaded-data" class="section level1">
<h1>Accessing the information in the loaded data</h1>
<p>The object created by <code>genDataRead</code> includes a lot of information. We have created functions that will help the user to navigate it.</p>
<div id="displaying-and-extracting-phenotype-information" class="section level2">
<h2>Displaying and extracting phenotype information</h2>
<p>First of all, when you type in the name of the object, a short summary will be displayed:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>my.gen.data</span></code></pre></div>
<pre><code>## This is raw genetic data read in through genDataRead.
## 
## It contains the following parts:
##    cov.data, gen.data, aux 
## 
## with following dimensions:
##   - covariate variables =  id.fam, id.c, id.f, id.m, sex, cc 
##       (total  6  covariate variables),
##   - number of markers =  429 ,
##   - number of data lines =  1659</code></pre>
<p>If you want to show and/or extract part of phenotype information, you can use the <code>showPheno</code> function:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="co"># by default - showing first 5 entries:</span></span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="kw">showPheno</span>( my.gen.data )</span></code></pre></div>
<pre><code>##      id.fam id.c id.f id.m sex cc 
## [1,] &quot;1&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [2,] &quot;1&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [3,] &quot;1&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;
## [4,] &quot;2&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;0&quot;
## [5,] &quot;2&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="co"># getting all the info:</span></span>
<span id="cb26-2"><a href="#cb26-2"></a><span class="kw">head</span>( <span class="kw">showPheno</span>( my.gen.data, <span class="dt">n =</span> <span class="st">&quot;all&quot;</span> ), <span class="dt">n =</span> <span class="dv">20</span> )</span></code></pre></div>
<pre><code>##       id.fam id.c id.f id.m sex cc 
##  [1,] &quot;1&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
##  [2,] &quot;1&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [3,] &quot;1&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;
##  [4,] &quot;2&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;0&quot;
##  [5,] &quot;2&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [6,] &quot;2&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;
##  [7,] &quot;3&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;0&quot;
##  [8,] &quot;3&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
##  [9,] &quot;3&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;0&quot;
## [10,] &quot;4&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;1&quot;
## [11,] &quot;4&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [12,] &quot;4&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;0&quot;
## [13,] &quot;5&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [14,] &quot;5&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [15,] &quot;5&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;
## [16,] &quot;6&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;1&quot;
## [17,] &quot;6&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [18,] &quot;6&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;0&quot;
## [19,] &quot;7&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [20,] &quot;7&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="kw">showPheno</span>( my.gen.data, <span class="dt">from =</span> <span class="dv">4</span>, <span class="dt">to =</span> <span class="dv">15</span> )</span></code></pre></div>
<pre><code>##       id.fam id.c id.f id.m sex cc 
##  [1,] &quot;2&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;0&quot;
##  [2,] &quot;2&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [3,] &quot;2&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;
##  [4,] &quot;3&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;0&quot;
##  [5,] &quot;3&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
##  [6,] &quot;3&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;0&quot;
##  [7,] &quot;4&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;1&quot; &quot;1&quot;
##  [8,] &quot;4&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [9,] &quot;4&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;0&quot;
## [10,] &quot;5&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [11,] &quot;5&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [12,] &quot;5&quot;    &quot;3&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot; &quot;1&quot;</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="co"># show information about females only:</span></span>
<span id="cb30-2"><a href="#cb30-2"></a><span class="kw">head</span>( <span class="kw">showPheno</span>( my.gen.data, <span class="dt">sex =</span> <span class="dv">2</span>, <span class="dt">n =</span> <span class="st">&quot;all&quot;</span> ), <span class="dt">n =</span> <span class="dv">20</span> )</span></code></pre></div>
<pre><code>##       id.fam id.c id.f id.m sex cc 
##  [1,] &quot;1&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
##  [2,] &quot;1&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [3,] &quot;2&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [4,] &quot;3&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
##  [5,] &quot;4&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [6,] &quot;5&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
##  [7,] &quot;5&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
##  [8,] &quot;6&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
##  [9,] &quot;7&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [10,] &quot;7&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [11,] &quot;8&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;0&quot;
## [12,] &quot;8&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [13,] &quot;9&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [14,] &quot;9&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [15,] &quot;10&quot;   &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [16,] &quot;11&quot;   &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [17,] &quot;11&quot;   &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [18,] &quot;12&quot;   &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [19,] &quot;13&quot;   &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [20,] &quot;14&quot;   &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;</code></pre>
<p>The output can be saved to an object:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>females.pheno &lt;-<span class="st"> </span><span class="kw">showPheno</span>( my.gen.data, <span class="dt">sex =</span> <span class="dv">2</span> )</span>
<span id="cb32-2"><a href="#cb32-2"></a><span class="kw">head</span>( females.pheno )</span></code></pre></div>
<pre><code>##      id.fam id.c id.f id.m sex cc 
## [1,] &quot;1&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;
## [2,] &quot;1&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [3,] &quot;2&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [4,] &quot;3&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;1&quot;
## [5,] &quot;4&quot;    &quot;2&quot;  &quot;0&quot;  &quot;0&quot;  &quot;2&quot; &quot;0&quot;
## [6,] &quot;5&quot;    &quot;1&quot;  &quot;3&quot;  &quot;2&quot;  &quot;2&quot; &quot;1&quot;</code></pre>
<p>With the functions <code>nindiv</code> and <code>nfam</code>, you can get the number of individuals or number of families in your data:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="kw">nindiv</span>( my.gen.data )</span></code></pre></div>
<pre><code>## [1] 1659</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="kw">nfam</span>( my.gen.data )</span></code></pre></div>
<pre><code>## [1] 550</code></pre>
<p>Note that the <code>nfam</code> function assumes that your dataset is from a triad or dyad study, i.e., includes information on the child and at least one parent.</p>
</div>
<div id="displaying-and-extracting-genotype-information" class="section level2">
<h2>Displaying and extracting genotype information</h2>
<p>The function <code>nsnps</code> will tell us how many markers/SNPs there is in the data. Be careful since per default it assumes that the data is triad data (i.e., mother, father, and child were genotyped), so if your data is from a case-control study, be sure to specify that with argument <code>design = &quot;cc&quot;</code>.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="kw">nsnps</span>( my.gen.data )</span></code></pre></div>
<pre><code>## [1] 429</code></pre>
<p>To get the SNP names use:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a><span class="kw">showSNPnames</span>( my.gen.data ) <span class="co"># by default - showing only first 5 SNPs</span></span></code></pre></div>
<pre><code>## [1] &quot;m1&quot; &quot;m2&quot; &quot;m3&quot; &quot;m4&quot; &quot;m5&quot;</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a><span class="kw">showSNPnames</span>( my.gen.data, <span class="dt">from =</span> <span class="dv">12</span>, <span class="dt">to =</span> <span class="dv">31</span> )</span></code></pre></div>
<pre><code>##  [1] &quot;m12&quot; &quot;m13&quot; &quot;m14&quot; &quot;m15&quot; &quot;m16&quot; &quot;m17&quot; &quot;m18&quot; &quot;m19&quot; &quot;m20&quot; &quot;m21&quot; &quot;m22&quot; &quot;m23&quot;
## [13] &quot;m24&quot; &quot;m25&quot; &quot;m26&quot; &quot;m27&quot; &quot;m28&quot; &quot;m29&quot; &quot;m30&quot; &quot;m31&quot;</code></pre>
<p>To extract and/or show genotypes for specific individuals or markers, use the <code>showGen</code> function:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a><span class="kw">showGen</span>( my.gen.data, <span class="dt">markers =</span> <span class="kw">c</span>( <span class="dv">10</span>,<span class="dv">15</span>,<span class="dv">121</span> ) ) <span class="co"># by default - showing first 5 entries</span></span></code></pre></div>
<pre><code>## opening ff C:/Users/hkgjess/AppData/Local/Temp/RtmpojjGlB/ff7ab0d214566.ff</code></pre>
<pre><code>## ff (open) byte length=30 (30) dim=c(5,6) dimorder=c(1,2) levels: G &lt;NA&gt; C T A
##      l_m10_a l_m10_b l_m15_a l_m15_b l_m121_a l_m121_b
## [1,]       T       T       A       T        T        T
## [2,]       T       T       T       T        T        T
## [3,]       T       T       A       T        T        T
## [4,]       A       T       A       T        T        T
## [5,]       A       T       T       T        T        T</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="kw">showGen</span>( my.gen.data, <span class="dt">from =</span> <span class="dv">31</span>, <span class="dt">to =</span> <span class="dv">231</span> )</span></code></pre></div>
<pre><code>## ff (open) byte length=2010 (2010) dim=c(201,10) dimorder=c(1,2) levels: G &lt;NA&gt; C T A
##        l_m1_a l_m1_b l_m2_a l_m2_b l_m3_a l_m3_b l_m4_a l_m4_b l_m5_a l_m5_b
## [1,]        G      G      T      A      T      T      G      G      G      G
## [2,]        G      G      A      A      T      T      G      G      G      G
## [3,]        G      G      T      A      T      T      G      G      G      G
## [4,]        G      G      A      A      A      T      G      G      C      G
## [5,]        G      G      T      A      A      T      G      G      C      G
## [6,]        G      G      A      A      A      T      G      G      C      G
## [7,]        G      G      A      A      A      T      G      G      G      G
## [8,]        G      G      A      A      A      T      G      G      G      G
## :           :      :      :      :      :      :      :      :      :      :
## [194,]      G      G      A      A      A      A      G      G      C      G
## [195,]      G      G      A      A      A      A      G      G      G      G
## [196,]      G      G      T      A      A      T      G      G      C      C
## [197,]      G      G      T      A      A      T      G      G      G      G
## [198,]      G      G      T      A      A      T      G      G      G      G
## [199,]      G      G      A      A      T      T      G      G      G      G
## [200,]      C      G      A      A      T      T      G      G      G      G
## [201,]      C      G      A      A      A      T      G      G      C      G</code></pre>
<p>As above, this output can be saved to an object:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a>subset.genes &lt;-<span class="st"> </span><span class="kw">showGen</span>( my.gen.data, <span class="dt">from =</span> <span class="dv">31</span>, <span class="dt">to =</span> <span class="dv">231</span>, <span class="dt">markers =</span> <span class="kw">c</span>( <span class="dv">10</span>,<span class="dv">15</span>,<span class="dv">121</span> ) )</span>
<span id="cb49-2"><a href="#cb49-2"></a>subset.genes</span></code></pre></div>
<pre><code>## ff (open) byte length=1206 (1206) dim=c(201,6) dimorder=c(1,2) levels: G &lt;NA&gt; C T A
##        l_m10_a l_m10_b l_m15_a l_m15_b l_m121_a l_m121_b
## [1,]        A       T       T       T        A        T 
## [2,]        A       T       T       T        T        T 
## [3,]        A       A       T       T        A        T 
## [4,]        T       T       T       T        A        T 
## [5,]        A       T       T       T        A        T 
## [6,]        T       T       T       T        A        T 
## [7,]        A       T       T       T        T        T 
## [8,]        A       T       T       T        T        T 
## :            :       :       :       :        :        :
## [194,]      A       T       T       T        T        T 
## [195,]      A       A       T       T        T        T 
## [196,]      T       T       T       T        T        T 
## [197,]      A       T       T       T        NA       NA
## [198,]      A       T       T       T        T        T 
## [199,]      A       A       T       T        A        T 
## [200,]      A       A       T       T        T        T 
## [201,]      A       A       T       T        T        T</code></pre>
<p><em>CAUTION:</em> Note that these functions work only with objects resulting from <code>genDataRead</code>, and not <code>genDataPreprocess</code>, as the preprocessing disturbs the coding of the data, thus making the output not easy to understand.</p>
</div>
</div>
<div id="preparing-your-data" class="section level1">
<h1>Preparing your data</h1>
<p>After loading the data, it is necessary to pre-process it to the internal format used by Haplin. This is done by evoking the command:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a>my.prepared.gen.data &lt;-<span class="st"> </span><span class="kw">genDataPreprocess</span>( <span class="dt">data.in =</span> my.gen.data, <span class="dt">map.file =</span></span>
<span id="cb51-2"><a href="#cb51-2"></a>  <span class="st">&quot;my_gen_data.map&quot;</span>, <span class="dt">design =</span> <span class="st">&quot;triad&quot;</span>, <span class="dt">file.out =</span> <span class="st">&quot;my_prepared_gen_data&quot;</span>,</span>
<span id="cb51-3"><a href="#cb51-3"></a>  <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span> )</span></code></pre></div>
<p><em>CAUTION:</em> This action can be very time-consuming for large datasets (e.g., estimated time for ca.45,000 SNPs and 1,600 individuals, a PED file of ca.700MB, is ca.6 minutes on a 8-core CPU). However, this needs to be done only once and the output, stored in small files, can be used for the subsequent analysis repeatedly. (See also section <a href="#subset">Choosing a subset of data</a> )</p>
<p>This will also create <code>.RData</code> and <code>.ffData</code> files, which take much less space than the input PED files. Be careful not to delete these files, as they can be re-used by simply loading into R (the <code>genDataLoad</code> function) right before Haplin analysis.</p>
<p><em>NOTE:</em> The information about the <code>my.prepared.gen.data</code> object can be displayed by simply writing the name of the object.</p>
<p>The output of the <code>genDataPreprocess</code> function (or <code>genDataLoad</code>) can then be used to run the analysis.</p>
</div>
<div id="subset" class="section level1">
<h1>Choosing a subset of data</h1>
<p>If you know that you want to focus your analysis on a certain region of the entire SNP set, or perhaps youâ€™re impatient and want to check out Haplin without waiting a long time for the preprocessing to finish, you can easily choose a subset of the data to pre-process and analyze. This can be done with the command:</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1"></a>gen.data.subset &lt;-<span class="st"> </span><span class="kw">genDataGetPart</span>( <span class="dt">data.in =</span> my.gen.data, <span class="dt">markers =</span> <span class="kw">c</span>( <span class="dv">3</span><span class="op">:</span><span class="dv">15</span>,<span class="dv">22</span> ),</span>
<span id="cb52-2"><a href="#cb52-2"></a>  <span class="dt">design =</span> <span class="st">&quot;triad&quot;</span>, <span class="dt">file.out =</span> <span class="st">&quot;my_gen_data_subset&quot;</span>, <span class="dt">dir.out =</span> <span class="st">&quot;.&quot;</span> )</span></code></pre></div>
<p>This function allows you to specify the subset in various ways:</p>
<ul>
<li><strong>markers</strong> - numeric vector with numbers indicating which markers to choose;</li>
<li><strong>indiv.ids</strong> - numeric vector giving IDs of individuals. <em>CAUTION:</em> in a standard PED file, individual IDs are not unique, so this will select all individuals with given numbers;</li>
<li><strong>rows</strong> - numeric vector giving the positions - this will select only these rows;</li>
<li><strong>cc</strong> - one or more values to choose based on the case-control status (â€˜ccâ€™ column);</li>
<li><strong>sex</strong> - one or more values to choose based on the â€˜sexâ€™ column;</li>
<li><strong>â€¦</strong> - if any additional covariate data are available in `data.in, the user can choose based on values of these.</li>
</ul>
<p>If you give a combination of these parameters, the result will be the intersection of the subsets defined by each of the parameters alone. The subset is then available in the <code>gen.data.subset</code> object and written to <code>file.out</code> file(s). These can be loaded and re-used multiple times.</p>
<div id="useful-shortcut-functions" class="section level2">
<h2>Useful shortcut functions</h2>
<p>To make it easier to choose certain subsets of data, we created the following functions:</p>
<ul>
<li><code>getChildren</code> - extracts only the childrenâ€™s data from the triad-design data;</li>
<li><code>getMothers</code> and <code>getFathers</code> - extracts only the fathers or mothers, respectively;</li>
<li><code>getFullTriads</code> - extracts only the full families, if the input data contains also dyads.</li>
</ul>
</div>
</div>
<div id="reuse" class="section level1">
<h1>Re-using the data</h1>
<p><em>IMPORTANT:</em> Remember that each time you start a new R session, you need to load the data into the memory with the command:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a>my.prepared.gen.data &lt;-<span class="st"> </span><span class="kw">genDataLoad</span>( <span class="dt">filename =</span> <span class="st">&quot;my_prepared_gen_data&quot;</span>,</span>
<span id="cb53-2"><a href="#cb53-2"></a>  <span class="dt">dir.in =</span> <span class="st">&quot;.&quot;</span> )</span></code></pre></div>
<p>This takes much less time than re-reading and running the data preparations!</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
