<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Miriam Gjerdevik and Hakon K. Gjessing" />

<meta name="date" content="2020-01-07" />

<title>Relative efficiency in Haplin</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Relative efficiency in Haplin</h1>
<h4 class="author">Miriam Gjerdevik and Hakon K. Gjessing</h4>
<h4 class="date">2020-01-07</h4>



<div id="background" class="section level2">
<h2>Background</h2>
<p>A variety of child-parent configurations are amenable to genetic association studies, including (but not limited to) cases in combination with unrelated controls, case-parent triads, and case-parent triads in combination with unrelated control-parent triads. Because genome-wide association studies (GWAS) are frequently underpowered due to the large number of single-nucleotide polymorphisms being tested, power calculations are necessary to choose an optimal study design and to maximize scientific gains from high genotyping and assay costs.</p>
<p>The statistical power is an important aspect of design comparison. Frequently, study designs are compared directly through a power analysis, without considering the total number of individuals that needs to be genotyped. For example, a fixed number of complete case-parent triads could be compared with the same number of case-mother or case-father dyads. However, such an approach ignores the costs of data collection. A much more general and informative design comparison can be achieved by studying the relative efficiency, which we define as the ratio of variances of two different parameter estimators, corresponding to two separate designs. Using log-linear modeling, we derive the relative efficiency from the asymptotic variance formulas of the parameters. The relative efficiency estimate takes into account the fact that different designs impose different costs relative to the number of genotyped individuals. The relative efficiency calculations are implemented as an easy-to-use function in our R package Haplin <span class="citation">(H.K. Gjessing and Lie 2006)</span>) .</p>
<p>We use the releative efficiency estimates to select the study design that attains the highest statistical power using the smallest sample collection and assay costs. The results will depend on the genetic effect being assessed, and our analyses include regular autosomal (offspring or child) effects, parent-of-origin effects and maternal effects (a definition of the genetic effects are provided in <span class="citation">(M. Gjerdevik et al. 2019)</span>). We here show example commands for various scenarios.</p>
</div>
<div id="regular-autosomal-effects" class="section level2">
<h2>Regular autosomal effects</h2>
<p>The relative efficiency of two designs are calculated by the Haplin function <code>hapRelEff</code>. The commands are very similar to the Haplin power calculation function <code>hapPowerAsymp</code>, which are explained in detail in our previously published paper <span class="citation">(M. Gjerdevik et al. 2019)</span>. In general, one only needs to specify the study designs to be compared, the allele frequencies, and the type of genetic effect and its magnitude.</p>
<p>The following command calculates the efficiency of the standard case-control design with an equal number of case and control children relative to the case-parent triad design.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">     <span class="kw">hapRelEff</span>(<span class="dt">cases.comp =</span> <span class="kw">c</span>(<span class="dt">c=</span><span class="dv">1</span>), 
            <span class="dt">controls.comp =</span> <span class="kw">c</span>(<span class="dt">c=</span><span class="dv">1</span>), <span class="dt">cases.ref =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>),
            <span class="dt">haplo.freq =</span> <span class="kw">c</span>(<span class="fl">0.1</span>,<span class="fl">0.9</span>), <span class="dt">RR =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
<pre><code>## $haplo.rel.eff
##   Haplotype RR.rel.eff
## 1         1        1.5
## 2         2        ref</code></pre>
<p>The arguments <code>cases.comp</code> and <code>controls.comp</code> specify the comparison designs, whereas <code>cases.ref</code> and <code>controls.ref</code> specify the reference design. We use the following abbreviations to describe the family designs. We let the letters c, m and f denote a child, mother and a father, respectively. Thus, the case-parent triad design is specified by <code>cases.comp = c(mfc=1)</code> or <code>cases.ref = c(mfc=1)</code>, whereas the standard case-control design is specified by <code>cases.comp = c(c=1)</code> and <code>controls.comp = c(c=1)</code> or <code>cases.ref = c(c=1)</code> and <code>controls.ref = c(c=1)</code>. To specify a case-control design with twice as many controls than cases, one could use the combination <code>cases.comp = c(c=1)</code> and <code>controls.comp = c(c=2)</code>.</p>
<p>The genetic effects are determined by the choice of relative risk parameter(s), which also specifies the effect sizes. A reguar autosomal effect is specified by the relative risk argument <code>RR</code>. The relative efficiency estimated under the null hypothesis, i.e., when all relative risks are equal to one, is known as the Pitman efficiency <span class="citation">(Noether 1955)</span>. However, other relative risk values can be used. Allele frequencies are specified by the argument <code>haplo.freq</code>. Note that the order and length of the specified relative risk parameter vectors should always match the corresponding allele frequencies.</p>
<p>We see that the relative efficiency for the standard case-control design is 1.5, compared with the case-parent triad design. This result is well-known from the literature <span class="citation">(H.J. Cordell and Clayton 2005)</span>.</p>
<p>To compare the full hybrid design consisting of both case-parent triads and control-parent triads, we can use a command similar to</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">     <span class="kw">hapRelEff</span>(<span class="dt">cases.comp =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>), 
            <span class="dt">controls.comp =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>), <span class="dt">cases.ref =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>),
            <span class="dt">haplo.freq =</span> <span class="kw">c</span>(<span class="fl">0.2</span>,<span class="fl">0.8</span>), <span class="dt">RR =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
</div>
<div id="parent-of-origin-poo-effects" class="section level2">
<h2>Parent-of-origin (PoO) effects</h2>
<p>The relative efficiency for PoO effects is computed by replacing the argument <code>RR</code> by the two relative risk arguments <code>RRcm</code> and <code>RRcf</code> denoting parental origin m (mother) and f (father). The command</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">     <span class="kw">hapRelEff</span>(<span class="dt">cases.comp =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>), 
            <span class="dt">controls.comp =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>), <span class="dt">cases.ref =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>),
            <span class="dt">haplo.freq =</span> <span class="kw">c</span>(<span class="fl">0.2</span>,<span class="fl">0.8</span>), <span class="dt">RRcm =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="dt">RRcf =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
<p>calculates the efficiency for the full hybrid design, relative to the case-parent triad design. We refer to our previous paper <span class="citation">(M. Gjerdevik et al. 2019)</span> for an explanation of the full output.</p>
</div>
<div id="maternal-effects" class="section level2">
<h2>Maternal effects</h2>
<p>Since children and their mothers have an allele in common, a maternal effect might be statistically confounded with a regular autosomal effect or a PoO effect. The relative efficiency for maternal effects can be analyzed jointly with that of a regular autosomal effect or a PoO effect by adding the relative risk argument <code>RR.mat</code> to the original command.</p>
<p>The command</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">     <span class="kw">hapRelEff</span>(<span class="dt">cases.comp =</span> <span class="kw">list</span>(<span class="kw">c</span>(<span class="dt">mc=</span><span class="dv">1</span>)), 
            <span class="dt">cases.ref=</span><span class="kw">list</span>(<span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>)), <span class="dt">haplo.freq =</span> <span class="kw">c</span>(<span class="fl">0.1</span>,<span class="fl">0.9</span>), 
            <span class="dt">RR =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="dt">RR.mat=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
<pre><code>## $haplo.rel.eff
##   Haplotype RR.rel.eff RRm.rel.eff
## 1         1        0.6         0.6
## 2         2        ref         ref</code></pre>
<p>calculates the efficiency of the case-mother dyad design relative to the case-mother dyad design, assessing both regular autosomal and maternal effects. In this example, we see that the relative efficiency estimates for regular autosomal and maternal effects are identical when adjusting for possible confounding of the effects with one another <span class="citation">(M. Gjerdevik et al. 2019)</span>.</p>
</div>
<div id="haplotype-effects" class="section level2">
<h2>Haplotype effects</h2>
<p>The default commands correspond to analyses of single-SNPs. However, the extention to haplotypes is straightforward. The number of markers and haplotypes is determined by the vector <code>nall</code>, where the number of markers is equal to <code>length(nall)</code>, and the number of different haplotypes is equal to <code>prod(nall)</code>. Thus, two diallelic markers are denoted by <code>nall = c(2,2)</code>. The length of the arguments <code>haplo.freq</code> and <code>RR</code> should correspond to the number of haplotypes, as shown in the example below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">     <span class="kw">hapRelEff</span>(<span class="dt">nall =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">cases.comp =</span> <span class="kw">c</span>(<span class="dt">c=</span><span class="dv">1</span>), 
            <span class="dt">controls.comp =</span> <span class="kw">c</span>(<span class="dt">c=</span><span class="dv">1</span>), <span class="dt">cases.ref =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>),
            <span class="dt">haplo.freq =</span> <span class="kw">c</span>(<span class="fl">0.1</span>,<span class="fl">0.2</span>,<span class="fl">0.3</span>,<span class="fl">0.4</span>), <span class="dt">RR =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
<pre><code>## $haplo.rel.eff
##   Haplotype RR.rel.eff
## 1       1-1       1.31
## 2       2-1       1.22
## 3       1-2       1.27
## 4       2-2        ref</code></pre>
<p>We recommend consulting our paper <span class="citation">(M. Gjerdevik et al. 2019)</span> for a more detailed description of haplotype analysis.</p>
</div>
<div id="analysis-of-x-linked-markers" class="section level2">
<h2>Analysis of X-linked markers</h2>
<p>Different X-chromosome models are implemented in Haplin, depending on the underlying assumptions of allele-effects in males versus females. The various models may include sex-specific baseline risks, common or distinct relative risks for males and females, as well as X-inactivation in females. Corresponding relative efficiency estimates are readily available in <code>hapRelEff</code>. In addition to the arguments needed to perform analyses on autosomal markers, three arguments must be specified for relative efficiency estimates on the X chromosome. First, to indicate an X-chromosome analysis, the argument <code>xchrom</code> must be set to <code>TRUE</code>. Second, the argument <code>sim.comb.sex</code> specifies how to deal with sex differences on the X-chromosome, i.e., X-inactivation or not. Finally, the argument <code>BR.girls</code> specifies the ratio of baseline risk for females relative to males. A detailed description of the parameterization models is provided elsewhere <span class="citation">(A. Jugessur et al. 2012; O. Skare et al. 2017; O. Skare et al. 2018)</span>.</p>
<p>The command</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">     <span class="kw">hapRelEff</span>(<span class="dt">cases.comp =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>), <span class="dt">controls.comp =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>), 
            <span class="dt">cases.ref =</span> <span class="kw">c</span>(<span class="dt">mfc=</span><span class="dv">1</span>),
            <span class="dt">haplo.freq =</span> <span class="kw">c</span>(<span class="fl">0.8</span>,<span class="fl">0.2</span>), 
            <span class="dt">RRcm =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">RRcf =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>),
            <span class="dt">xchrom =</span> T, <span class="dt">sim.comb.sex =</span> <span class="st">&quot;double&quot;</span>,
            <span class="dt">BR.girls =</span> <span class="dv">1</span>)</code></pre></div>
<p>estimates the PoO relative efficiency for the full hybrid design versus the case-parent triad design, accounting for X-inactivation in females (<code>sim.comb.sex = &quot;double&quot;</code>) and assuming the same baseline risk in females and males (<code>BR.girls = 1</code>). We refer to our previously published paper <span class="citation">(M. Gjerdevik et al. 2019)</span> for further details.</p>
<hr />
<div id="refs" class="references">
<div id="ref-Jugessur2012a">
<p>A. Jugessur, O. Skare, R.T. Lie, A.J. Wilcox, K. Christensen, L. Christiansen, T.T. Nguyen, J. C. Murray, and H.K. Gjessing. 2012. “X-linked genes and risk of orofacial clefts: evidence from two population-based studies in Scandinavia.” <em>PLoS One</em> 7 (6): e39240.</p>
</div>
<div id="ref-Cordell2005">
<p>H.J. Cordell, and D.G. Clayton. 2005. “Genetic association studies.” <em>Lancet</em> 366 (9491): 1121–31.</p>
</div>
<div id="ref-Gjessing2006a">
<p>H.K. Gjessing, and R.T. Lie. 2006. “Case-parent triads: estimating single- and double-dose effects of fetal and maternal disease gene haplotypes.” <em>Ann. Hum. Genet.</em> 70 (3): 382–96.</p>
</div>
<div id="ref-Gjerdevik2019">
<p>M. Gjerdevik, A. Jugessur, O.A. Haaland, J. Romanowska, R.T. Lie, H.J. Cordell, and H.K. Gjessing. 2019. “Haplin power analysis: a software module for power and sample size calculations in genetic association analyses of family triads and unrelated controls.” <em>BMC Bioinformatics</em> 20 (1): 165.</p>
</div>
<div id="ref-Noether1955">
<p>Noether, G.E. 1955. “On a theorem of Pitman.” <em>Ann. Math. Stat.</em> 26 (1): 64–68.</p>
</div>
<div id="ref-Skare2017">
<p>O. Skare, H.K. Gjessing, M. Gjerdevik, O.A. Haaland, J. Romanowska, R.T. Lie, and A. Jugessur. 2017. “A new approach to chromosome-wide analysis of X-linked markers identifies new associations in Asian and European case-parent triads of orofacial clefts.” <em>PLoS One</em> 12 (9): e0183772.</p>
</div>
<div id="ref-Skare2018">
<p>O. Skare, R.T. Lie, O.A. Haaland, M. Gjerdevik, J. Romanowska, H.K. Gjessing, and A. Jugessur. 2018. “Analysis of parent-of-origin effects on the X chromosome in Asian and European orofacial cleft triads identifies associations with <em>DMD</em>, <em>FGF13</em>, <em>EGFL6</em>, and additional loci at Xp22.2.” <em>Front. Genet.</em> 9 (25). doi:<a href="https://doi.org/10.3389/fgene.2018.00025">10.3389/fgene.2018.00025</a>.</p>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
