<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Miriam Gjerdevik and Hakon K. Gjessing" />

<meta name="date" content="2022-05-19" />

<title>Relative efficiency in Haplin</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Relative efficiency in Haplin</h1>
<h4 class="author">Miriam Gjerdevik and Hakon K. Gjessing</h4>
<h4 class="date">2022-05-19</h4>



<div id="background" class="section level2">
<h2>Background</h2>
<p>A variety of child-parent configurations are amenable to genetic
association studies, including (but not limited to) cases in combination
with unrelated controls, case-parent triads, and case-parent triads in
combination with unrelated control-parent triads. Because genome-wide
association studies (GWAS) are frequently underpowered due to the large
number of single-nucleotide polymorphisms being tested, power
calculations are necessary to choose an optimal study design and to
maximize scientific gains from high genotyping and assay costs.</p>
<p>The statistical power is an important aspect of design comparison.
Frequently, study designs are compared directly through a power
analysis, without considering the total number of individuals that needs
to be genotyped. For example, a fixed number of complete case-parent
triads could be compared with the same number of case-mother or
case-father dyads. However, such an approach ignores the costs of data
collection. A much more general and informative design comparison can be
achieved by studying the relative efficiency, which we define as the
ratio of variances of two different parameter estimators, corresponding
to two separate designs. Using log-linear modeling, we derive the
relative efficiency from the asymptotic variance formulas of the
parameters. The relative efficiency estimate takes into account the fact
that different designs impose different costs relative to the number of
genotyped individuals. The relative efficiency calculations are
implemented as an easy-to-use function in our R package Haplin <span class="citation">(H. K. Gjessing and Lie 2006)</span>) .</p>
<p>We use the releative efficiency estimates to select the study design
that attains the highest statistical power using the smallest sample
collection and assay costs. The results will depend on the genetic
effect being assessed, and our analyses include regular autosomal
(offspring or child) effects, parent-of-origin effects and maternal
effects (a definition of the genetic effects are provided in <span class="citation">(M. Gjerdevik et al. 2019)</span>). We here show
example commands for various scenarios.</p>
</div>
<div id="regular-autosomal-effects" class="section level2">
<h2>Regular autosomal effects</h2>
<p>The relative efficiency of two designs are calculated by the Haplin
function <code>hapRelEff</code>. The commands are very similar to the
Haplin power calculation function <code>hapPowerAsymp</code>, which are
explained in detail in our previously published paper <span class="citation">(M. Gjerdevik et al. 2019)</span>. In general, one only
needs to specify the study designs to be compared, the allele
frequencies, and the type of genetic effect and its magnitude.</p>
<p>The following command calculates the efficiency of the standard
case-control design with an equal number of case and control children
relative to the case-parent triad design.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>     <span class="fu">hapRelEff</span>(<span class="at">cases.comp =</span> <span class="fu">c</span>(<span class="at">c=</span><span class="dv">1</span>), </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">controls.comp =</span> <span class="fu">c</span>(<span class="at">c=</span><span class="dv">1</span>), <span class="at">cases.ref =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>),</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">haplo.freq =</span> <span class="fu">c</span>(<span class="fl">0.1</span>,<span class="fl">0.9</span>), <span class="at">RR =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code></pre></div>
<pre><code>## $haplo.rel.eff
##   Haplotype RR.rel.eff
## 1         1        1.5
## 2         2        ref</code></pre>
<p>The arguments <code>cases.comp</code> and <code>controls.comp</code>
specify the comparison designs, whereas <code>cases.ref</code> and
<code>controls.ref</code> specify the reference design. We use the
following abbreviations to describe the family designs. We let the
letters c, m and f denote a child, mother and a father, respectively.
Thus, the case-parent triad design is specified by
<code>cases.comp = c(mfc=1)</code> or <code>cases.ref = c(mfc=1)</code>,
whereas the standard case-control design is specified by
<code>cases.comp = c(c=1)</code> and <code>controls.comp = c(c=1)</code>
or <code>cases.ref = c(c=1)</code> and
<code>controls.ref = c(c=1)</code>. To specify a case-control design
with twice as many controls than cases, one could use the combination
<code>cases.comp = c(c=1)</code> and
<code>controls.comp = c(c=2)</code>.</p>
<p>The genetic effects are determined by the choice of relative risk
parameter(s), which also specifies the effect sizes. A reguar autosomal
effect is specified by the relative risk argument <code>RR</code>. The
relative efficiency estimated under the null hypothesis, i.e., when all
relative risks are equal to one, is known as the Pitman efficiency <span class="citation">(Noether 1955)</span>. However, other relative risk
values can be used. Allele frequencies are specified by the argument
<code>haplo.freq</code>. Note that the order and length of the specified
relative risk parameter vectors should always match the corresponding
allele frequencies.</p>
<p>We see that the relative efficiency for the standard case-control
design is 1.5, compared with the case-parent triad design. This result
is well-known from the literature <span class="citation">(H. J. Cordell
and Clayton 2005)</span>.</p>
<p>To compare the full hybrid design consisting of both case-parent
triads and control-parent triads, we can use a command similar to</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>     <span class="fu">hapRelEff</span>(<span class="at">cases.comp =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>), </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">controls.comp =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>), <span class="at">cases.ref =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">haplo.freq =</span> <span class="fu">c</span>(<span class="fl">0.2</span>,<span class="fl">0.8</span>), <span class="at">RR =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code></pre></div>
</div>
<div id="parent-of-origin-poo-effects" class="section level2">
<h2>Parent-of-origin (PoO) effects</h2>
<p>The relative efficiency for PoO effects is computed by replacing the
argument <code>RR</code> by the two relative risk arguments
<code>RRcm</code> and <code>RRcf</code> denoting parental origin m
(mother) and f (father). The command</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>     <span class="fu">hapRelEff</span>(<span class="at">cases.comp =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>), </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">controls.comp =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>), <span class="at">cases.ref =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>),</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">haplo.freq =</span> <span class="fu">c</span>(<span class="fl">0.2</span>,<span class="fl">0.8</span>), <span class="at">RRcm =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">RRcf =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code></pre></div>
<p>calculates the efficiency for the full hybrid design, relative to the
case-parent triad design. We refer to our previous paper <span class="citation">(M. Gjerdevik et al. 2019)</span> for an explanation of
the full output.</p>
</div>
<div id="maternal-effects" class="section level2">
<h2>Maternal effects</h2>
<p>Since children and their mothers have an allele in common, a maternal
effect might be statistically confounded with a regular autosomal effect
or a PoO effect. The relative efficiency for maternal effects can be
analyzed jointly with that of a regular autosomal effect or a PoO effect
by adding the relative risk argument <code>RR.mat</code> to the original
command.</p>
<p>The command</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>     <span class="fu">hapRelEff</span>(<span class="at">cases.comp =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="at">mc=</span><span class="dv">1</span>)), </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">cases.ref=</span><span class="fu">list</span>(<span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>)), <span class="at">haplo.freq =</span> <span class="fu">c</span>(<span class="fl">0.1</span>,<span class="fl">0.9</span>), </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">RR =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">RR.mat=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code></pre></div>
<pre><code>## $haplo.rel.eff
##   Haplotype RR.rel.eff RRm.rel.eff
## 1         1        0.6         0.6
## 2         2        ref         ref</code></pre>
<p>calculates the efficiency of the case-mother dyad design relative to
the case-mother dyad design, assessing both regular autosomal and
maternal effects. In this example, we see that the relative efficiency
estimates for regular autosomal and maternal effects are identical when
adjusting for possible confounding of the effects with one another <span class="citation">(M. Gjerdevik et al. 2019)</span>.</p>
</div>
<div id="haplotype-effects" class="section level2">
<h2>Haplotype effects</h2>
<p>The default commands correspond to analyses of single-SNPs. However,
the extention to haplotypes is straightforward. The number of markers
and haplotypes is determined by the vector <code>nall</code>, where the
number of markers is equal to <code>length(nall)</code>, and the number
of different haplotypes is equal to <code>prod(nall)</code>. Thus, two
diallelic markers are denoted by <code>nall = c(2,2)</code>. The length
of the arguments <code>haplo.freq</code> and <code>RR</code> should
correspond to the number of haplotypes, as shown in the example
below.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>     <span class="fu">hapRelEff</span>(<span class="at">nall =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="at">cases.comp =</span> <span class="fu">c</span>(<span class="at">c=</span><span class="dv">1</span>), </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">controls.comp =</span> <span class="fu">c</span>(<span class="at">c=</span><span class="dv">1</span>), <span class="at">cases.ref =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>),</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">haplo.freq =</span> <span class="fu">c</span>(<span class="fl">0.1</span>,<span class="fl">0.2</span>,<span class="fl">0.3</span>,<span class="fl">0.4</span>), <span class="at">RR =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))</span></code></pre></div>
<pre><code>## $haplo.rel.eff
##   Haplotype RR.rel.eff
## 1       1-1       1.31
## 2       2-1       1.22
## 3       1-2       1.27
## 4       2-2        ref</code></pre>
<p>We recommend consulting our paper <span class="citation">(M.
Gjerdevik et al. 2019)</span> for a more detailed description of
haplotype analysis.</p>
</div>
<div id="analysis-of-x-linked-markers" class="section level2">
<h2>Analysis of X-linked markers</h2>
<p>Different X-chromosome models are implemented in Haplin, depending on
the underlying assumptions of allele-effects in males versus females.
The various models may include sex-specific baseline risks, common or
distinct relative risks for males and females, as well as X-inactivation
in females. Corresponding relative efficiency estimates are readily
available in <code>hapRelEff</code>. In addition to the arguments needed
to perform analyses on autosomal markers, three arguments must be
specified for relative efficiency estimates on the X chromosome. First,
to indicate an X-chromosome analysis, the argument <code>xchrom</code>
must be set to <code>TRUE</code>. Second, the argument
<code>sim.comb.sex</code> specifies how to deal with sex differences on
the X-chromosome, i.e., X-inactivation or not. Finally, the argument
<code>BR.girls</code> specifies the ratio of baseline risk for females
relative to males. A detailed description of the parameterization models
is provided elsewhere <span class="citation">(A. Jugessur et al. 2012;
O. Skare et al. 2017, 2018)</span>.</p>
<p>The command</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>     <span class="fu">hapRelEff</span>(<span class="at">cases.comp =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>), <span class="at">controls.comp =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>), </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">cases.ref =</span> <span class="fu">c</span>(<span class="at">mfc=</span><span class="dv">1</span>),</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">haplo.freq =</span> <span class="fu">c</span>(<span class="fl">0.8</span>,<span class="fl">0.2</span>), </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">RRcm =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">RRcf =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>),</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">xchrom =</span> T, <span class="at">sim.comb.sex =</span> <span class="st">&quot;double&quot;</span>,</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">BR.girls =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>estimates the PoO relative efficiency for the full hybrid design
versus the case-parent triad design, accounting for X-inactivation in
females (<code>sim.comb.sex = &quot;double&quot;</code>) and assuming the same
baseline risk in females and males (<code>BR.girls = 1</code>). We refer
to our previously published paper <span class="citation">(M. Gjerdevik
et al. 2019)</span> for further details.</p>
<hr />
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Jugessur2012a" class="csl-entry">
A. Jugessur, O. Skare, R. T. Lie, A. J. Wilcox, K. Christensen, L.
Christiansen, T. T. Nguyen, J. C. Murray, and H. K. Gjessing. 2012.
<span>“<span class="nocase">X-linked genes and risk of orofacial clefts:
evidence from two population-based studies in
Scandinavia</span>.”</span> <em>PLoS One</em> 7 (6): e39240.
</div>
<div id="ref-Cordell2005" class="csl-entry">
H. J. Cordell, and D. G. Clayton. 2005. <span>“<span class="nocase">Genetic association studies</span>.”</span>
<em>Lancet</em> 366 (9491): 1121–31.
</div>
<div id="ref-Gjessing2006a" class="csl-entry">
H. K. Gjessing, and R. T. Lie. 2006. <span>“<span class="nocase">Case-parent triads: estimating single- and double-dose
effects of fetal and maternal disease gene haplotypes</span>.”</span>
<em>Ann. Hum. Genet.</em> 70 (3): 382–96.
</div>
<div id="ref-Gjerdevik2019" class="csl-entry">
M. Gjerdevik, A. Jugessur, O. A. Haaland, J. Romanowska, R. T. Lie, H.
J. Cordell, and H. K. Gjessing. 2019. <span>“<span class="nocase">Haplin
power analysis: a software module for power and sample size calculations
in genetic association analyses of family triads and unrelated
controls</span>.”</span> <em>BMC Bioinformatics</em> 20 (1): 165.
</div>
<div id="ref-Noether1955" class="csl-entry">
Noether, G. E. 1955. <span>“<span class="nocase">On a theorem of
Pitman</span>.”</span> <em>Ann. Math. Stat.</em> 26 (1): 64–68.
</div>
<div id="ref-Skare2017" class="csl-entry">
O. Skare, H. K. Gjessing, M. Gjerdevik, O. A. Haaland, J. Romanowska, R.
T. Lie, and A. Jugessur. 2017. <span>“<span class="nocase">A new
approach to chromosome-wide analysis of X-linked markers identifies new
associations in Asian and European case-parent triads of orofacial
clefts</span>.”</span> <em>PLoS One</em> 12 (9): e0183772.
</div>
<div id="ref-Skare2018" class="csl-entry">
O. Skare, R. T. Lie, O. A. Haaland, M. Gjerdevik, J. Romanowska, H. K.
Gjessing, and A. Jugessur. 2018. <span>“<span class="nocase">Analysis of
parent-of-origin effects on the X chromosome in Asian and European
orofacial cleft triads identifies associations with <em>DMD</em>,
<em>FGF13</em>, <em>EGFL6</em>, and additional loci at
Xp22.2</span>.”</span> <em>Front. Genet.</em> 9 (25). <a href="https://doi.org/10.3389/fgene.2018.00025">https://doi.org/10.3389/fgene.2018.00025</a>.
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
